### PR TITLE
Codes to use DM Exposures as inputs

### PR DESCRIPTION
**Pull request by esheldon**
These are parallel versions with DM appended to the names.

They will replace the old ones once a full shear recover test has been passed

### CODE REVIEW

**Comment by beckermr** [comment]
I will get to this next week.

**Comment by esheldon** [comment]
@erykoff I've added a skeleton for online coadds (see above). Untested as of now.

File: descwl_coadd/online_coadds.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+
+    mask_map = AssembleCoaddTask.setRejectedMaskMapping(stats_ctrl)
+
+    return AccumulatorMeanStack(
+        shape=dims,
+        bit_mask_value=stats_ctrl.getAndMask(),
+        mask_threshold_dict={},
+        mask_map=mask_map,
+        no_good_pixels_mask=stats_ctrl.getNoGoodPixelsMask(),
+        calc_error_from_input_variance=True,
```
Comment (erykoff) [inline descwl_coadd/online_coadds.py:97]: This should be `stats_ctrl.getCalcErrorFromInputVariance()`

File: descwl_coadd/online_coadds.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+        The coadd bbox
+    """
+
+    # dims is C/numpy ordering
+    dims = get_dims_from_bbox(coadd_bbox)
+
+    mask_map = AssembleCoaddTask.setRejectedMaskMapping(stats_ctrl)
+
+    return AccumulatorMeanStack(
+        shape=dims,
```
Comment (erykoff) [inline descwl_coadd/online_coadds.py:92]: I'm not sure why you do the whole bbox dance; if you send the coadd down the pipe you can get the array dimensions from there without worrying about the row/column flipping.

File: descwl_coadd/online_coadds.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+            exp,
+            maxBBox=exp.getBBox(),
+            destBBox=coadd_bbox,
+        )
+
+        stacker.add_masked_image(wexp.image, weight=weight)
+
+        # if isinstance(texp, DeferredDatasetHandle):
+        #     # is this needed?
+        #     del exp
```
Comment (erykoff) [inline descwl_coadd/online_coadds.py:67]: Should not be needed, it goes out of scope and will be garbage collected.

File: descwl_coadd/online_coadds.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+        Bits for which the pixels will not be added to the coadd.
+        e.g. we would not let EDGE pixels get coadded
+
+    Returns
+    -------
+    afw_math.StatisticsControl
+    """
+    stats_ctrl = afw_math.StatisticsControl()
+    stats_ctrl.setAndMask(mask)
+    stats_ctrl.setWeighted(True)
```
Comment (erykoff) [inline descwl_coadd/online_coadds.py:118]: This isn't used by the online code at all.

File: descwl_coadd/online_coadds.py [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+    # dims is C/numpy ordering
+    dims = get_dims_from_bbox(coadd_bbox)
+
+    mask_map = AssembleCoaddTask.setRejectedMaskMapping(stats_ctrl)
+
+    return AccumulatorMeanStack(
+        shape=dims,
+        bit_mask_value=stats_ctrl.getAndMask(),
+        mask_threshold_dict={},
```
Comment (erykoff) [inline descwl_coadd/online_coadds.py:94]: If you make this dict blank then (currently) any bad pixels (edge?) won't get marked in the coadd at all.  Maybe this is what you want!

**Review by erykoff — COMMENTED** [review]
I've added some comments to the online draft.

File: descwl_coadd/online_coadds.py (hunk 5 above)
Comment (esheldon) [reply descwl_coadd/online_coadds.py:94]: so why is the andMask set then?

Comment (erykoff) [reply descwl_coadd/online_coadds.py:94]: So the settings here will mask any bits in the from `stats_ctrl.getAndMask()` but it won't let you know in the coadd mask plane where those bits were masked.

Comment (esheldon) [reply descwl_coadd/online_coadds.py:94]: I see; what do you recommend I send for mask_threshold_dict then?

Comment (esheldon) [reply descwl_coadd/online_coadds.py:94]: Currently I've only got EDGE in the `andMask` but that could change

Comment (erykoff) [reply descwl_coadd/online_coadds.py]: So if you want to know wherever there was a bad edge input (I think you do!) then the threshold should be 0 for that bit.  But it depends on what your threshold is!  Any? 5%? 10% of the input images?  Etc.

File: descwl_coadd/online_coadds.py [hunk 6]
```diff
@@ -83,18 +81,18 @@ def make_online_stacker(stats_ctrl, coadd_bbox):
...
     """
 
-    # dims is C/numpy ordering
-    dims = get_dims_from_bbox(coadd_bbox)
-
     mask_map = AssembleCoaddTask.setRejectedMaskMapping(stats_ctrl)
 
+    cefiv = stats_ctrl.getCalcErrorFromInputVariance()
+
+    mask_threshold_dict = stats_ctrl_to_threshold_dict(stats_ctrl)
```
Comment (esheldon) [inline descwl_coadd/online_coadds.py:91]: @erykoff I added this, which I suppose is just using defaults

For EDGE at least we don't want *any* to make it through, so how do I make that happen?  Sorry for so many detailed questions.

**Comment by esheldon** [comment]
going to  merge even though online coadds not completely tested. For longer tests I like to make a tag and  @beckermr would prefer tags only off master

**Pull request merged**